Optimize HttpMessageHandlerIntegration

kevingosse opened the pull request:
- Make SpanContextPropagator.Inject generic to prevent the boxing of the headers collection
- Add an overload in TypeExtensions to avoid building the full type name
- Stop using String keys in AsyncHelper
- Reuse the boxed cancellationtoken when needed
- Split SendAsyncInternal to avoid allocating the async state machine on the fast path
- Cache the "namespace and name filters" array
- Cache the return task type

**Benchmark:**

```csharp
public class CustomMessageHandler : HttpMessageHandler
{
	private static Task<HttpResponseMessage> _cachedResult = Task.FromResult(new HttpResponseMessage());
	private static HttpResponseMessage _cachedResponse = new HttpResponseMessage();

	protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
	{
		return _cachedResult;
	}
}

(Task)HttpMessageHandlerIntegration.HttpMessageHandler_SendAsync(...).Wait();
```

FastPath: Takes the "skip instrumentation" path in `SendAsyncInternal`
Full: Goes through the instrumentation code

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.19041
Intel Core i7-9750H CPU 2.60GHz, 1 CPU, 12 logical and 6 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4180.0), X64 RyuJIT
  Job-VTGAFA : .NET Framework 4.8 (4.8.4180.0), X64 RyuJIT
  Job-MNUBEA : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT


```

**.NET 4.7.2:**

|                    Method |          Mean |         Error |         StdDev |        Median |  Gen 0 |  Gen 1 |  Gen 2 | Allocated |
|-------------------------- |--------------:|--------------:|---------------:|--------------:|-------:|-------:|-------:|----------:|
| SendRequest_NoIntegration |      3.781 ns |     0.1032 ns |      0.1447 ns |      3.706 ns |      - |      - |      - |         - |
|      SendRequest_Full_Old | 18,710.139 ns |   399.8844 ns |    460.5075 ns | 18,551.764 ns | 3.2043 | 0.4578 | 0.1526 |   20201 B |
|      SendRequest_Full_New | 11,077.879 ns |   121.0923 ns |    107.3451 ns | 11,080.820 ns | 0.4120 | 0.1984 |      - |    2624 B |
|  SendRequest_Fastpath_Old |  6,020.003 ns |    94.4310 ns |     88.3308 ns |  5,999.944 ns | 2.8534 | 0.0153 |      - |   17991 B |
|  SendRequest_Fastpath_New |  2,177.383 ns |    42.6974 ns |     37.8501 ns |  2,162.498 ns | 0.0725 | 0.0076 |      - |     457 B |


**.NET Core 3.1:**

|                    Method |          Mean |         Error |         StdDev |        Median |  Gen 0 |  Gen 1 |  Gen 2 | Allocated |
|-------------------------- |--------------:|--------------:|---------------:|--------------:|-------:|-------:|-------:|----------:|
| SendRequest_NoIntegration |      4.122 ns |     0.0380 ns |      0.0337 ns |      4.116 ns |      - |      - |      - |         - |
|      SendRequest_Full_Old | 15,527.897 ns |   304.9692 ns |    501.0731 ns | 15,470.544 ns | 3.4485 | 0.5798 | 0.1526 |   21319 B |
|      SendRequest_Full_New |  7,810.876 ns |   151.9542 ns |    134.7034 ns |  7,782.911 ns | 0.3815 | 0.1831 |      - |    2416 B |
|  SendRequest_Fastpath_Old |  5,331.927 ns |   105.6334 ns |    187.7632 ns |  5,289.986 ns | 3.0823 | 0.0153 |      - |   19359 B |
|  SendRequest_Fastpath_New |  1,132.209 ns |     6.6264 ns |      5.5334 ns |  1,131.997 ns | 0.0725 |      - |      - |     456 B |

kevingosse commented:
It's already a nice improvement but I believe there is still room for more. I'll iterate on this, either in this PR or in a separate one.

lucaspimentel commented on src/Datadog.Trace.ClrProfiler.Managed.Core/TypeExtensions.cs:
```diff
@@ -26,6 +26,31 @@ internal static class TypeExtensions
+
+            while (currentType != null)
+            {
+                if (currentType.Namespace == instrumentedNamespace && currentType.Name == instrumentedTypeName)
```
Would comparing the type name first before the namespace be a little faster in the common case? This is probably very tiny optimization, but also zero cost.

Two reasons: types can be in the same namespace, which in the current code means two string comparisons instead of one. Additionally, namespaces tend to be longer and start with a common substring. For example, when comparing `System.Data.SqlClient.SqlCommand` and `System.Data.Common.DbCommand`, the runtime can determine they are not equal as soon as it looks at the first character in the type names (`S` vs `D`), but it takes a little longer to compare the namespaces because they both start with `System.Data.`.

kevingosse replied:
Sounds reasonable, done

zacharycmontoya replied:
Did this get implemented? It still looks like the namespace is tested first

lucaspimentel replied:
I flipped them in 8c39f42d

kevingosse commented:
In the end I stopped relying on AsyncHelper and added an extension method to convert a task type to another. It makes the code much easier to follow (in my opinion).

I also moved the "no instrumentation" path earlier in the method to reduce the overhead.

Final results:

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.19041
Intel Core i7-9750H CPU 2.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.302
  [Host]     : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT
  Job-UDBQJC : .NET Framework 4.8 (4.8.4180.0), X64 RyuJIT
  Job-UEXFKP : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT


```
|                   Method |       Runtime |        Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------------------------- |-------------- |------------:|----------:|----------:|-------:|-------:|------:|----------:|
|     SendRequest_Full_Old |    .NET 4.7.2 | 14,343.3 ns | 410.78 ns | 403.44 ns | 3.1738 |      - |     - |   20115 B |
|     SendRequest_Full_New |    .NET 4.7.2 |  9,191.7 ns |  76.87 ns |  68.14 ns | 0.3815 |      - |     - |    2423 B |
| SendRequest_Fastpath_Old |    .NET 4.7.2 |  5,445.0 ns |  31.44 ns |  29.41 ns | 2.8687 | 0.0153 |     - |   18072 B |
| SendRequest_Fastpath_New |    .NET 4.7.2 |  1,321.7 ns |  15.28 ns |  13.54 ns | 0.0343 |      - |     - |     217 B |
|     SendRequest_Full_Old | .NET Core 3.1 | 12,221.1 ns |  67.97 ns |  60.26 ns | 3.3875 | 0.0153 |     - |   21295 B |
|     SendRequest_Full_New | .NET Core 3.1 |  6,074.7 ns |  56.31 ns |  49.91 ns | 0.3510 |      - |     - |    2224 B |
| SendRequest_Fastpath_Old | .NET Core 3.1 |  5,439.5 ns | 107.52 ns |  95.31 ns | 3.0899 | 0.0153 |     - |   19431 B |
| SendRequest_Fastpath_New | .NET Core 3.1 |    563.3 ns |   9.63 ns |   8.54 ns | 0.0343 |      - |     - |     216 B |

lucaspimentel commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/HttpMessageHandlerIntegration.cs:
```diff
@@ -19,11 +18,18 @@ public static class HttpMessageHandlerIntegration
         private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.GetLogger(typeof(HttpMessageHandlerIntegration));
+        private static readonly string[] NamespaceAndNameFilters = { ClrNames.GenericTask, ClrNames.HttpRequestMessage, ClrNames.CancellationToken };
+
+        private static Type _httpMessageHandlerResultType;
+        private static Type _httpClientHandlerResultType;
```
There may be an edge case here where we end up with multiple versions of `HttpMessageHandler` or `HttpClientHandler` loaded into the app domain. Just to make sure I understand, is this what you're checking for below in `taskResultType.Assembly != httpMessageHandler.Assembly`?

kevingosse replied:
I preferred to play safe yes. Worst case scenario this will revert the optimization and we'll be back to fetching the type every time

tonyredondo commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/HttpMessageHandlerIntegration.cs:
```diff
@@ -231,8 +254,10 @@ public static class HttpMessageHandlerIntegration
+                    var task = (Task)sendAsync(handler, request, cancellationToken);
+                    await task.ConfigureAwait(false);
+
+                    var response = task.GetProperty("Result").Value;
```
Maybe we can avoid the Lookup (and maybe allocations) inside GetProperty here by declaring a static var with a copy of the delegate compiled from the expression tree to extract the result. If we already know what is the type of T, like `HttpResponseMessage` we can pass the delegate as a parameter to SendAsyncInternal. So the caller pass the delegate to extract the value of the task.

kevingosse replied:
`GetProperty` does not allocate. You're right that it would save a lookup, but I think the overhead is negligible with the current state of the method (see the measured overhead of GetProperty here: https://github.com/DataDog/dd-trace-dotnet/pull/759). I think that's something we should keep for another round of optimizations.

tonyredondo replied:
Ok 👍
